AtomNode tests

RyanR712 opened the pull request:
Comprehensive tests for AtomNode

Commit:
AtomNode.test.ts, misc changes
settings.json - updated spellcheck
AtomNode.ts - empty string error checking, changed toString() format, deleted containsNode() and cutNode Ellipse imports
Ellipse.test.ts - points contained on the Ellipse's boundaries should now be considered contained
Rectangle.test.ts - points contained on the Rectangle's boundaries should now be considered contained
AtomNode.test.ts - constructor, contains and toString() tests

Commit:
Merge branch 'master' into 124-atom-node-tests

AnushaTiwari5 committed:
contains vs intersects logic

AnushaTiwari5 commented on src/AEG/AEGTree.ts:
```diff
@@ -108,6 +108,40 @@ export class AEGTree {
+     * @param otherNode the other node
+     * @returns the result of each shape's respective intersect() methods.
+     */
+    private intersects(incomingNode: AtomNode | CutNode, otherNode: AtomNode | CutNode) {
+        if (incomingNode instanceof AtomNode) {
+            if (otherNode instanceof AtomNode) {
+                return shapesIntersect(
+                    (incomingNode as AtomNode).rectangle,
+                    (otherNode as AtomNode).rectangle
+                );
+            } else {
+                return shapesIntersect(
+                    (incomingNode as AtomNode).rectangle,
+                    (otherNode as CutNode).ellipse!
+                );
+            }
+        } else {
+            if (otherNode instanceof AtomNode) {
+                return shapesIntersect(
+                    (incomingNode as CutNode).ellipse!,
+                    (otherNode as AtomNode).rectangle
+                );
+            } else {
+                return shapesIntersect(
+                    (incomingNode as CutNode).ellipse!,
+                    (otherNode as CutNode).ellipse!
+                );
+            }
+        }
+    }
```
we have been using ! because ellipse can return null, should we find a different workaround?

AnushaTiwari5 committed:
Update src/AEG/AEGTree.ts

minimize intersects()

Co-authored-by: James Oswald <jamesoswald111@gmail.com>

AnushaTiwari5 committed:
minimize AEGTree intersects(), no more default constructors

Commit:
AtomNode tests complete, default constructors gone
Point.test.ts - removed default constructor tests
Ellipse.test.ts - removed default constructor tests, moved some tests to other test blocks
Rectangle.test.ts - moved test cases to better reflect expected function behavior during testing
AtomNode.ts - added length and Latin alphabet checks for the identifier
AtomNode.test.ts - comprehensive implementation!
All tests pass!

Commit:
Minor toString() change in AtomNode
AtomNode.ts - made the toString more accurate. Comprehensiveness not compromised.

RyanR712 commented:
I meant error message in this previous commit, not toString

James-Oswald commented on src/AEG/AtomNode.ts:
```diff
@@ -26,13 +24,28 @@ export class AtomNode {
+                    " passed in as identifier in AtomNode constructor, which is not of length 1."
+            );
+        }
+        if (new RegExp(/^[A-Za-z]$/).test(val) === false) {
```
Can we run test directly on the regex without the wrapper new RegExp wrapper? Isn't it already a regex object?

James-Oswald commented on src/AEG/AEGUtils.ts:
```diff
@@ -11,22 +11,26 @@ import {Rectangle} from "./Rectangle";
+    return shapesIntersect(newShape, existingShape) || shapeContains(existingShape, newShape);
+}
+
+export function shapesIntersect(
```
Needs a doc comment.

James-Oswald commented on src/AEG/AEGUtils.ts:
```diff
@@ -11,22 +11,26 @@ import {Rectangle} from "./Rectangle";
-            return ellipseRectangleCollision(newShape as Ellipse, existingShape as Rectangle);
+            return ellipseRectangleIntersection(newShape as Ellipse, existingShape as Rectangle);
         } else {
-            //For ellipse-ellipse collision, check if the rectangular bounding boxes intersect.
-            //If they do, check if points of the new ellipse are within the current ellipse
```
You should leave this comment, this is very useful information as it wouldn't be clear to most people why you're running this check

James-Oswald commented on src/AEG/AEGUtils.ts:
```diff
@@ -154,22 +172,23 @@ function edgesWithin(rect1: Rectangle, rect2: Rectangle): boolean {
  * @param ellipse The given ellipse
  * @param otherPoint The point that might be inside the given ellipse.
- * @returns True, if the point is inside the given ellipse. Else, false
+ * @returns 1, if the point is on the ellipse.
+ * Returns <1 if the point is completely within the ellipse.
```
You're missing a case; this can also be greater than 1 if the point is outside the ellipse.

James-Oswald commented on src/AEG/AEGUtils.ts:
```diff
@@ -154,22 +172,23 @@ function edgesWithin(rect1: Rectangle, rect2: Rectangle): boolean {
 
 /**
- * Method that checks whether a point is within the given ellipse.
+ * Method that returns the value of a point compared to the ellipse
```
"Value compared to a point" is not a clear explanation of what this value represents, my suggestion is to rework this to be "signedDistanceFromEllipse" by moving the one to the other side of the equation:
 ```
 (x-h)^2/rx^2 + (y-k)^2/ry^2 <= 1
 (x-h)^2/rx^2 + (y-k)^2/ry^2 - 1 <= 0
  ```
 Now this is a signed distance metric, equal to 0 means the point is on the ellipse (no displacement). <0 means inside the ellipse and >0 means outside the ellipse. 

Note that this is not a distance metric since distance is always positive, signed distance is directed and is actually based off of the topological notion of a point being inside of something! Perfect for our application.

https://en.wikipedia.org/wiki/Signed_distance_function

James-Oswald replied:
Albeit a true signed distance function is positive if the point is inside and negative outside

AnushaTiwari5 committed:
resolved reviewed comments, fixed ellipse point division by 0
